netsim/modules/__init__.py

```diff
@@ -458,38 +461,41 @@ def get_effective_module_attribute(
 
 mod_load: typing.Dict = {}
 
-def module_transform(method: str, topology: Box) -> None:
+def call_module_hook(module: str, method: str, topology: Box, **kwargs: typing.Any) -> None:
   global mod_load
 
+  if not mod_load.get(module):
+    mod_load[module] = _Module.load(module,topology.get(module))
+  if log.debug_active('modules'):
+    if hasattr(mod_load[module],method):
+      print(f'Calling module {module} {method}')
+      if kwargs:
+        print(f'... args: {kwargs}')
+  mod_load[module].call(method,topology=topology,**kwargs)
+
+def execute_module_hooks(method: str, topology: Box) -> None:
+  for m in list_of_modules(topology):
+    if method in topology.defaults[m].get('hooks',[]):
+      call_module_hook(m,f'module_{method}',topology)
+
+def module_transform(method: str, topology: Box) -> None:
   if log.debug_active('modules'):
     print(f'Processing module_{method} hooks')
 
-  for m in topology.get('module',[]) + topology.get('_extra_module',[]):
-    if not mod_load.get(m):
-      mod_load[m] = _Module.load(m,topology.get(m))
-    if log.debug_active('modules'):
-      if hasattr(mod_load[m],f"module_{method}"):
-        print(f'Calling module {m} module_{method}')
-    mod_load[m].call("module_"+method,topology)
+  m_list = topology.get('module',[])
+  m_list += [ m for m in topology.get('_extra_module',[]) if m not in m_list ]
+  for m in m_list:
```

Comment:
This assumes there is no need to check dependency order for 'hooks'?

Reply:
Exactly. However, the `_extra_module` ones are added to the module list in `merge_global_module_params`(before pre_transform hooks are called), so we should be good.

This change was focused more on ensuring a module listed in `_extra_module` (current use case: **bgp.policy** plugin adding **routing** module) is not called twice.